### DISCUSSION TITLE
Configure Mod key

### DISCUSSION THREAD

**Discussion by xiro-codes** [YaLTeR/niri#197] [Ideas]
// "Mod" is a special modifier equal to Super when running on a TTY, and to Alt
    // when running as a winit window. 

Why...., I literally need the reverse behavior

**Comment by YaLTeR** [comment]
Super is generally viewed as a key for compositor / windowing system. Most applications won't have any default binds that include Super, making it a natural choice as the TTY Mod key.

On the contrary, Alt frequently appears in default bindings in apps, and also it is the standard key for mnemonics in UI toolkits. This is when you hold Alt in GTK apps, and underlines appear on one letter of every button or menu row, and you can press Alt + that letter to activate it.

**Reply by JeffreyLeeLi** [reply]
Using Super as default Mod key is a good choice, in most cases, but I encountered 2 specific problems.

I run Fedora in VirtualBox, and:
#1: When host is macOS, Super+H will hide entire guest window.
#2: When host is Windows, Super+L is the default key to lock the system.

I tried to find ways to deactivate os's behavior, but no luck for now.
Also, cannot see relations with the inhibiting behavior.

So, hope you can help me to figure out a way out, thanks in advance.

P.S. I want to say: Niri is so cool and scrollable tiling is super fantastic. I cannot stop toying with it at first sight, although it has been only a couple of days.

**Reply by JeffreyLeeLi** [reply]
I finally got what I expected using config from: https://variety4me.github.io/niri_docs/Configuration-Input/#mod-key-mod-key-nested

For anyone who needs:
```
input {
    mod-key "Alt"
    mod-key-nested "Super"
}
```

**Reply by Sempyos** [reply]
It's also on the official niri wiki here: https://niri-wm.github.io/niri/Configuration%3A-Input.html#mod-key-mod-key-nested

**Reply by JeffreyLeeLi** [reply]
Thanks @Sempyos

**Comment by bittin** [comment]
would be nice if you could configure the keys, had problems with running Manjaro awesome in a gnome-boxes where mod + r was eaten by Niri

**Reply by YaLTeR** [reply]
For this one we just need to implement the shortcut inhibit portal / protocol, then it'll work.

**Reply by bittin** [reply]
then i request that to be done ;)

**Reply by YaLTeR** [reply]
Made an issue: https://github.com/YaLTeR/niri/issues/198

**Reply by markstos** [reply]
The inhibit protocol has now been added: https://github.com/YaLTeR/niri/pull/630

**Comment by ghost** [comment]
Just in case less experienced users get confused when arriving to this thread: you can still change Mod (the function) to Alt by changing the "Mod" keyword to "Alt" in the config file (I've done it and works fine). You may need to change some other bindings as "Alt+Alt+something" is not valid. 

This discussion is about being able to configure the "Mod" keyword to mean "Alt" or something else.

**Comment by Quackdoc** [comment]
I have a somewhat related question, Instead of just rebinding Mod, I would like to also be able to define my own

I use some custom input devices that have hard coded buttons, Reading the docs it seems like Niri only supports the below for modifiers. I would like to be able to set something like `Mod = Menu`  `Mod1 = XF86HomePage`. I could rebind the device inside of linux with something like input-remapper for Mod, but ideally this wouldn't be needed, and it wouldn't help the second mod situation.

**Reply by YaLTeR** [reply]
There's a bit of a problem here in that niri can currently only use modifiers that XKB thinks are modifiers (so only the usual ones like Ctrl, Shift, Alt, Super, and also ISO Level3/5 Shift). So you might be better off rebinding via XKB.

**Comment by dantefromhell** [comment]
I'm using NixOS which has the awesome feature of being able to boot a system declaration in a qemu-based VM for testing.

Due to the way niri works though the keyboard shortcuts inside the VM and outside are the same, meaning none of my shortcuts work inside the VM.

I stumbled over

> When running on a TTY, the Mod key is Super. When running in a window, the Mod key is Alt.

It would be great if there's a niri config option to force this behaviour as NixOS allows to configure options that get applied to a system only when it's booted in a VM.

That would allow me to have the normal keyboard shortcuts for my host system but force them to be 'Alt' based when booting the VM.

**Reply by YaLTeR** [reply]
FWIW this PR implements the shortcuts inhibit protocol that should make this unnecessary: #630

**Reply by dantefromhell** [reply]
This is awesome, lan-mouse is on my TODO list to get working w/ niri to finally make for a better experience commanding my RPIs! @sodiboo Huge thanks for your work, I'll be peaking/ stealing some of your nix code right away ;-)

**Reply by sodiboo** [reply]
> > When running on a TTY, the Mod key is Super. When running in a window, the Mod key is Alt.
> 
> It would be great if there's a niri config option to force this behaviour as NixOS allows to configure options that get applied to a system only when it's booted in a VM.
> 
> That would allow me to have the normal keyboard shortcuts for my host system but force them to be 'Alt' based when booting the VM.

Though niri doesn't currently have this capability, you can already solve your use case in a different way! Just tell niri to always make them `Alt` iff you're in a VM. [See my config, which already does the *exact* thing you describe](https://github.com/sodiboo/system/blob/b6795a6ae9ae06fa487e413cbbb23b696e018e66/niri.mod.nix#L131-L139)

**Reply by dantefromhell** [reply]
Turns out using 'Alt' as Mod key collides with other keyboard shortcuts using 'Alt'.
Is there a way to use 'Alt_R' as a mod key instead?

**Comment by mgutz** [comment]
I highly recommend [keyd](https://github.com/rvaiya/keyd) for all your keyboard remapping needs. keyd can differentiate between tap and hold, so for example you can configure your super key to send super on hold and super+space on tap. Remapping any key is straightforward.
